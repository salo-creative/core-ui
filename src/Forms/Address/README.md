# Address

This is a compound component designed to simplify Address captures within applications. It captures address details in a standardised format. The field requires the first line of address, city and postcode. Beyond this you can also capture line 2 of address, county and country although these fields are not required. 

It is also possible to manage which of these additional fields are shown by passing an array of field names on the fields prop. If a blank array is passed then only the 3 required fields will render.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Address from '@salo/core-ui/Forms/Address';
```

Implement as follows

```javascript
<Address
  name='field_name'
  value={ this.state.address }
  onChange={ address => this.setState({  address }) }
/>
```