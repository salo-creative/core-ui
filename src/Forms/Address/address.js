import React from 'react';
import PropTypes from 'prop-types';
import { string } from 'yup';

// COMPONENTS & STYLES
import Input from '../Input';

// HELPERS & CONSTANTS
import { colours } from '../../helpers/colours';

const allowedFields = ['line2', 'county', 'country'];

const Address = (props) => {
  const {
    background,
    border,
    borderRadius,
    className,
    disabled,
    error,
    fields,
    fontSize,
    margin,
    name,
    onChange,
    padding,
    required,
    size,
    value
  } = props;

  // Hold state of individual inputs within component
  const [state, setState] = React.useState({
    line1: '',
    line2: '',
    city: '',
    county: '',
    postcode: '',
    country: '',
    ...value
  });

  // On field blur we emit these back on the on change function to the parent
  const handleBlur = () => {
    onChange(state);
  };

  // This is needed to trigger field validation when return is pressed to submit
  const handleKeyUp = ({ e }) => {
    if (e.keyCode === 13) {
      handleBlur();
    }
  };

  const handleChange = ({ key, val }) => {
    setState({
      ...state,
      [key]: val
    });
  };

  const standardProps = {
    background,
    border,
    borderRadius,
    disabled,
    fontSize,
    margin,
    padding,
    size,
    onBlur: handleBlur,
    onKeyUp: handleKeyUp
  };

  return (
    <div className={ `address__wrapper ${ className }` }>
      <Input
        { ...standardProps } // eslint-disable-line react/jsx-props-no-spreading
        error={ error && !string().required().isValidSync(state.line1) }
        errorMessage='The first line of address is required'
        key={ `line1_${ name }` }
        label='Address line 1'
        name={ `line1_${ name }` }
        required={ required }
        value={ state.line1 }
        onChange={ ({ value: val }) => handleChange({ key: 'line1', val }) }
      />
      { fields.includes('line2') && (
        <Input
          { ...standardProps } // eslint-disable-line react/jsx-props-no-spreading
          error={ error && !string().isValidSync(state.line2) }
          key={ `line2_${ name }` }
          label='Address line 2'
          name={ `line2_${ name }` }
          value={ state.line2 }
          onChange={ ({ value: val }) => handleChange({ key: 'line2', val }) }
        />
      ) }
      <Input
        { ...standardProps } // eslint-disable-line react/jsx-props-no-spreading
        error={ error && !string().required().isValidSync(state.city) }
        errorMessage='A city is required'
        key={ `city_${ name }` }
        label='City'
        name={ `city_${ name }` }
        required={ required }
        value={ state.city }
        onChange={ ({ value: val }) => handleChange({ key: 'city', val }) }
      />
      { fields.includes('county') && (
        <Input
          { ...standardProps } // eslint-disable-line react/jsx-props-no-spreading
          error={ error && !string().isValidSync(state.county) }
          key={ `county_${ name }` }
          label='County'
          name={ `county_${ name }` }
          value={ state.county }
          onChange={ ({ value: val }) => handleChange({ key: 'county', val }) }
        />
      ) }
      <Input
        { ...standardProps } // eslint-disable-line react/jsx-props-no-spreading
        error={ error && !string().required().isValidSync(state.postcode) }
        errorMessage='A valid postcode is required'
        key={ `postcode_${ name }` }
        label='Postcode'
        name={ `postcode_${ name }` }
        required={ required }
        value={ state.postcode }
        onChange={ ({ value: val }) => handleChange({ key: 'postcode', val }) }
      />
      { fields.includes('country') && (
        <Input
          { ...standardProps } // eslint-disable-line react/jsx-props-no-spreading
          error={ error && !string().isValidSync(state.country) }
          key={ `country_${ name }` }
          label='Country'
          name={ `country_${ name }` }
          value={ state.country }
          onChange={ ({ value: val }) => handleChange({ key: 'country', val }) }
        />
      ) }
    </div>
  );
};

Address.defaultProps = {
  background: colours.paleGrey,
  border: '1px solid',
  borderRadius: '0.4rem',
  className: '',
  disabled: false,
  error: false,
  fields: allowedFields,
  fontSize: '1.4rem',
  margin: '0 0 2rem',
  onChange: () => null,
  padding: '0 1rem',
  required: false,
  size: 'M',
  value: {}
};

Address.propTypes = {
  background: PropTypes.string,
  border: PropTypes.string,
  borderRadius: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  fields: PropTypes.arrayOf(PropTypes.oneOf(allowedFields)),
  fontSize: PropTypes.string,
  margin: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  padding: PropTypes.string,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M']),
  value: PropTypes.object
};

export default Address;