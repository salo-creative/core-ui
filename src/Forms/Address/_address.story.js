import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, select } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Address } from '../../index';

// README //
import README from './README.md';

// Start of story logic
const stories = storiesOf('Forms | Address', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['address'] });

stories.add(
  'Basic',
  (() => {
    // STORE
    const store = new Store({ value: {} });
    // KNOBS
    const padding = text('Padding', '0 1rem');
    const border = text('Border', '1px solid');
    const borderRadius = text('Border radius', '0.4rem');
    const fontSize = text('Font size', '1.4rem');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const size = select('size', ['L', 'M'], 'M');
    return (
      <RenderWithProps store={ store }>
        <Address
          name='story'
          disabled={ disabled }
          error={ error }
          fontSize={ fontSize }
          required={ required }
          size={ size }
          padding={ padding }
          border={ border }
          borderRadius={ borderRadius }
          value={ store.value }
          onChange={ (address) => console.log(address) }
        />
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);