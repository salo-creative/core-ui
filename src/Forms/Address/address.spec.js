import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Address from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Address />
  );
  expect(container.firstChild).toMatchSnapshot();
});