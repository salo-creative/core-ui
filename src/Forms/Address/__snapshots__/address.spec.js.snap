// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0 0 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  position: relative;
}

.c3 {
  margin: 0;
  padding: 0 1rem;
  height: 4rem;
  background-color: #fafcff;
  border: 1px solid;
  border-color: rgba(0,172,237,0.5);
  -webkit-transition: border 0.2s linear;
  transition: border 0.2s linear;
  width: 100%;
  border-radius: 0.4rem;
  font-size: 1.4rem;
}

.c3:focus,
.c3:active {
  border-color: #00ACED;
}

.c3[disabled] {
  cursor: not-allowed;
  opacity: 0.75;
}

.c1 {
  font-size: 1.4rem;
  margin: 0 0 0.5rem;
  padding: 0;
  color: inherit;
  word-break: break-word;
}

<div
  class="address__wrapper "
>
  <div
    class="c0 salo-input "
  >
    <label
      class="c1 salo-input__label"
      for="line1_undefined"
    >
      Address line 1
    </label>
    <div
      class="c2 salo-input__wrapper"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="c3 salo-input__field"
        font-size="1.4rem"
        height="4rem"
        id="line1_undefined"
        name="line1_undefined"
        placeholder="Enter address line 1"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="c0 salo-input "
  >
    <label
      class="c1 salo-input__label"
      for="line2_undefined"
    >
      Address line 2
    </label>
    <div
      class="c2 salo-input__wrapper"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="c3 salo-input__field"
        font-size="1.4rem"
        height="4rem"
        id="line2_undefined"
        name="line2_undefined"
        placeholder="Enter address line 2"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="c0 salo-input "
  >
    <label
      class="c1 salo-input__label"
      for="city_undefined"
    >
      City
    </label>
    <div
      class="c2 salo-input__wrapper"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="c3 salo-input__field"
        font-size="1.4rem"
        height="4rem"
        id="city_undefined"
        name="city_undefined"
        placeholder="Enter a city"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="c0 salo-input "
  >
    <label
      class="c1 salo-input__label"
      for="county_undefined"
    >
      County
    </label>
    <div
      class="c2 salo-input__wrapper"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="c3 salo-input__field"
        font-size="1.4rem"
        height="4rem"
        id="county_undefined"
        name="county_undefined"
        placeholder="Enter a county"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="c0 salo-input "
  >
    <label
      class="c1 salo-input__label"
      for="postcode_undefined"
    >
      Postcode
    </label>
    <div
      class="c2 salo-input__wrapper"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="c3 salo-input__field"
        font-size="1.4rem"
        height="4rem"
        id="postcode_undefined"
        name="postcode_undefined"
        placeholder="Enter a postcode"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="c0 salo-input "
  >
    <label
      class="c1 salo-input__label"
      for="country_undefined"
    >
      Country
    </label>
    <div
      class="c2 salo-input__wrapper"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="c3 salo-input__field"
        font-size="1.4rem"
        height="4rem"
        id="country_undefined"
        name="country_undefined"
        placeholder="Enter a country"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
